### ISSUE TITLE
Document setup procedure and decisions (including decisions so far)

### ISSUE THREAD

**Issue by shankari** [shankari/everest-demo#1]
High level goals are in the README

They already have an `automated-ci-testing-base-image` and `
`automated-ci-testing-work-image`. What do those include?

There is a `.ci` directory which has
- buildkit: `compile.sh` and `test_and_install.sh`
- e2e: which composes with MQTT

There is a `.github/workflows` which
- seems to run `everest/everest-ci/github-actions/run-clang-format@v1.0.0`
- and some kind of build-kit container
- `ghcr.io/everest/build-kit-alpine:latest`
- and then manually runs build/test/install
- and then runs MQTT with the other modules for the integration tests

**Comment by shankari** [comment]
So it seems like we should be able to create a new Dockerfile that builds on build-kit-alpine, compiles to create a new pre-built image, and then launches a basic SIL

**Comment by shankari** [comment]
First check:
- where does the build-kit-alpine come from? If there is already a pre-built image, we don't need to reinvent the wheel
- probably the everest-ci. Bingo. There is the build-kit but nothing else.
- where do automated-ci-testing... come from then?
- there is only one reference to it in the entire repo and that is the package, which is from May 9
https://github.com/search?q=org%3AEVerest+automated-ci-testing-base-image&type=registrypackages
- No other references to "automated" in https://github.com/search?q=org%3AEVerest+automated&type=code
- Ah found the reference in a PR: https://github.com/EVerest/everest-core/pull/228/files which was closed, not merged, and the files don't exist
- The PR was closed on May 24th and the images are from May 9th, so they were almost certainly generated from that brief moment in time.

**Comment by shankari** [comment]
(I guess we can ask them what happened, for now, we can potentially use that PR as a template of how to build)
Is that more recent or is the buildkit image build more recent?
(or did this morph into the buildkit image?)

**Comment by shankari** [comment]
Bingo! They created the new file on May 13th and updated on May 14th. Last update to the PR was May 14th
So basically they decided to move it over to a separate repo. The PR is obsolete. The new github action is the correct one and I can move forward with it.

**Comment by shankari** [comment]
Ok, so now that this has percolated into my brain, I think I know what I want to do.
I want to make a repo called everest-demo (currently running in my personal org)
It has the Dockerfile that I write and the docker-compose and the PWD

Depending on our final decision, we either submit this to EVerest as is, or
do the GitHub commands to merge it in without commit history (easier) or with
commit history (I've done it before).

**Comment by shankari** [comment]
(this is that repo). All updates before this were copied from a file I was maintaining.
Moving to this issue now.

**Comment by shankari** [comment]
The build-kit action is only manually triggered (on `workflow_dispatch`). Was successfully built 5 days ago
In `everest-core`, however, things are more murky.
There is a single `build_and_test` workflow defined in the repo, whose stability seems to depend on the particular PR
It doesn't look like they have any scheduled tasks runs defined to assess long-term stability

However, they *do* have a couple of other actions that are apparently run on pull requests as well
https://github.com/EVerest/everest-core/actions/workflows/build_ci_testing_base_image.yml
https://github.com/EVerest/everest-core/actions/workflows/run_ci_testing.yml

However, neither of them have run for 3 months so it looks like they used to have other workflows but removed them?
Although I don't see that in the history for the file, unless they deleted the directory and recreated.

**Comment by shankari** [comment]
Final question before starting this.
How/when do we rebuild the docker image for the demo and where do we store it?
Options:
- automated build
   - schedule github workflow task to run every n days
   - push image up if successful
   - have docker-compose use "latest"
- manual build
   - do the same steps but manually
   - manually bump up the image number on docker-compose
   - docker-compose up to test
   - push the image manually
   - PWD can use pushed image directly

That makes it fairly clear - start with manual, particularly for a demo!!
Can automate based on workflow run later. But no nasty demo surprises due to automated update!

Fortunately, we can push images to the GitHub docker registry with a manual token
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry

Fortunately, docker-compose also supports pushing the generated images
https://docs.docker.com/compose/compose-file/build/#publishing-built-images

```
Usage:  docker compose [OPTIONS] COMMAND
Commands:
  push        Push service images
```

I do think that we need to be careful about the name in the docker-compose - it might need to start with `ghcr.io/NAMESPACE` and we need to configure docker-compose accordingly.

`docker push ghcr.io/NAMESPACE/IMAGE_NAME:latest`

Enough procrastination/investigation - time to start with the code and open a PR
Detailed discussion will now move to the PR

**Comment by shankari** [comment]
Another issue is that the current workflow runs clang-format through a GitHub action (<a href="https://github.com/EVerest/everest-ci/blob/main/github-actions/run-clang-format/action.yaml">run-clang-format</a>) which in turn, runs the action as a docker command

```
runs:
  using: 'docker'
  image: 'docker://ghcr.io/everest/everest-clang-format:latest'
  args:
    - /github/workspace/${{ inputs.source-dir }}
    - --extensions
    - ${{ inputs.extensions }}
    - -e
    - /github/workspace/${{ inputs.source-dir }}/${{ inputs.exclude }}
    - --color
    - ${{ inputs.color }}
    - -r
```

I don't think I can run a docker command while building a docker image.

Options are:
(1) Try and find where that image is being built from and copy _its_ Dockerfile
(2) Look at the manual instructions and copy them over

I would really like to use the same process that the CI does, to increase the chances that this will work, so I don't want to use (2) unless I have to.

I wonder if we can use `docker-compose` dependencies to work around this - but I think that they only work for launch, not for build.

**Comment by shankari** [comment]
Hm, maybe we can use docker-compose order to get around this
If I create a `docker-compose` based on `.ci/e2e/docker-compose.yaml`

```
services:
  mqtt-server:
    build: mosquitto

  e2e-test-server:
    build: demo
    depends_on:
      - mqtt-server
```

<details>
<summary>They are built in order mosquitto -> demo</summary>

```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 104B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/eclipse-mosquitto:2.0.10
#3 ...

#4 [auth] library/eclipse-mosquitto:pull token for registry-1.docker.io
#4 DONE 0.0s

#3 [internal] load metadata for docker.io/library/eclipse-mosquitto:2.0.10
#3 DONE 2.8s

#5 [1/2] FROM docker.io/library/eclipse-mosquitto:2.0.10@sha256:683189aaaf01240fe4ba3c0b3262704c4a3910c7293162e22e8f43a319fc9ed8
#5 resolve docker.io/library/eclipse-mosquitto:2.0.10@sha256:683189aaaf01240fe4ba3c0b3262704c4a3910c7293162e22e8f43a319fc9ed8 0.0s done
#5 sha256:683189aaaf01240fe4ba3c0b3262704c4a3910c7293162e22e8f43a319fc9ed8 1.41kB / 1.41kB done
#5 sha256:7221c4efbc1ca0715f680e9f98fa467364c9ee26969963cb80052f54ac2804e5 946B / 946B done
#5 sha256:7e47c4a289482d2e875e3c499782fa065cea7c6c8928014fed812918bbc42e3e 8.81kB / 8.81kB done
#5 sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b 0B / 2.80MB 0.1s
#5 sha256:8cc7b833e2db4000ca2d4fea3477e90fe3c6bad4c3f34574514bb43c4a12fc3c 0B / 1.89MB 0.1s
#5 sha256:f2b454175e7e7e2efd42dbeaefbc17f7f515eb71fb8ccebd50449a89c1d9caba 0B / 368B 0.1s
#5 sha256:f2b454175e7e7e2efd42dbeaefbc17f7f515eb71fb8ccebd50449a89c1d9caba 368B / 368B 0.7s done
#5 sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b 1.05MB / 2.80MB 1.0s
#5 sha256:8cc7b833e2db4000ca2d4fea3477e90fe3c6bad4c3f34574514bb43c4a12fc3c 1.05MB / 1.89MB 1.3s
#5 sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b 2.10MB / 2.80MB 2.0s
#5 sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b 2.80MB / 2.80MB 2.3s done
#5 sha256:8cc7b833e2db4000ca2d4fea3477e90fe3c6bad4c3f34574514bb43c4a12fc3c 1.89MB / 1.89MB 2.2s done
#5 extracting sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b
#5 extracting sha256:339de151aab4bc06eed8409daae147c408478cb538dacb90cc63f19ad4eba80b 0.1s done
#5 extracting sha256:8cc7b833e2db4000ca2d4fea3477e90fe3c6bad4c3f34574514bb43c4a12fc3c 0.1s done
#5 extracting sha256:f2b454175e7e7e2efd42dbeaefbc17f7f515eb71fb8ccebd50449a89c1d9caba done
#5 DONE 2.7s

#6 [2/2] RUN echo "Building demo MQTT image"
#6 0.187 Building demo MQTT image
#6 DONE 0.2s

#7 exporting to image
#7 exporting layers done
#7 writing image sha256:d8049fb3a722facaceb7e8b20d642870f88c80d8689668903188ae6e70017c38 done
#7 naming to docker.io/library/everest-demo-mqtt-server done
#7 DONE 0.0s
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 121B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for ghcr.io/everest/build-kit-alpine:latest
#3 DONE 0.8s

#4 [1/2] FROM ghcr.io/everest/build-kit-alpine:latest@sha256:8c2841b885f22fb6e0d4733a8c980da370138b4f0d7ff364ba437d1ed9316089
#4 CACHED

#5 [2/2] RUN echo "Building demo docker image"
#5 0.194 Building demo docker image
#5 DONE 0.2s

#6 exporting to image
#6 exporting layers 0.0s done
#6 writing image sha256:0650f3dcd5e1df1dafa209c9042a79beac0a7102fc0167cff81c868874494f67 done
#6 naming to docker.io/library/everest-demo-e2e-test-server done
#6 DONE 0.0s
```

</details>

But if I switch the order in `docker-compose`, they are still built in the same order

<details>
<summary>They are built in order mosquitto -> demo</summary>

```
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 31B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/eclipse-mosquitto:2.0.10
#3 DONE 0.7s

#4 [1/2] FROM docker.io/library/eclipse-mosquitto:2.0.10@sha256:683189aaaf01240fe4ba3c0b3262704c4a3910c7293162e22e8f43a319fc9ed8
#4 CACHED

#5 [2/2] RUN echo "Building demo MQTT image"
#5 0.180 Building demo MQTT image
#5 DONE 0.2s

#6 exporting to image
#6 exporting layers 0.0s done
#6 writing image sha256:30add2daf408e6641b014d214232f79a3599c8955d9a1b79a6f7e28e71c6db07 done
#6 naming to docker.io/library/everest-demo-mqtt-server
#6 naming to docker.io/library/everest-demo-mqtt-server done
#6 DONE 0.0s
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 31B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for ghcr.io/everest/build-kit-alpine:latest
#3 DONE 0.4s

#4 [1/2] FROM ghcr.io/everest/build-kit-alpine:latest@sha256:8c2841b885f22fb6e0d4733a8c980da370138b4f0d7ff364ba437d1ed9316089
#4 CACHED

#5 [2/2] RUN echo "Building demo docker image"
#5 0.224 Building demo docker image
#5 DONE 0.2s

#6 exporting to image
#6 exporting layers 0.0s done
#6 writing image sha256:a957d967055ceecf44307dc3b8fa9d0e3ea7f76b47d0623b7e0e2ecbaa2c7c5c done
#6 naming to docker.io/library/everest-demo-e2e-test-server done
#6 DONE 0.0s
```
</details>

**Comment by shankari** [comment]
I also found the Dockerfile for the clang container
https://github.com/EVerest/everest-utils/blob/main/docker/everest-clang-format/Dockerfile

And now I understand the reason for the different containers - the buildkit container is based on alpine while the clang one is based on debian.

But there is a debian dockerfile as well
https://github.com/EVerest/everest-ci/tree/main/docker/images/build-kit

But they don't build and release debian images

So is this just clang not working on debian? Let's create a container and find out

**Comment by shankari** [comment]
Trying to follow the clang `Dockerfile` instructions, `curl https://github.com/EVerest/run-clang-format/releases/download/clang-format-16/clang-format` doesn't seem to return anything.

But downloading it from https://github.com/EVerest/run-clang-format/releases/download/clang-format-16/ does seem to work

I wonder if this is restricted to logged in users, which is why we need the ssh keys? But this is a docker command, so I don't see why it should matter.

Verified that it is downloadable using a private window
Or if I put it into a docker container

Moving on to trying to run it.

```
/ext # /usr/bin/run-clang-format
usage: run-clang-format [-h] [--clang-format-executable EXECUTABLE] [--extensions EXTENSIONS] [-r] [-d] [-i]
                        [-q] [-j N] [--color {auto,always,never}] [-e PATTERN] [--style STYLE]
                        file [file ...]
```

**Comment by shankari** [comment]
I have cloned the core repo and moved it to the `ext` directory to be consistent with the buildkit

```
          docker run \
          --volume "$(pwd):/ext" \
          --name compile-container \
          build-kit run-script compile
```

But the action is called with the arguments 

```
          source-dir: source
          extensions: hpp,cpp
          exclude: cache
```

And I don't find a `source` dir in the core.
Doh! it is reconfigured in the setup actions

```
      - name: Checkout everest-core
        uses: actions/checkout@v3
        with:
          path: source
```

So the "source" directory is the one where the `.git` repo is, so `ext` for us

**Comment by shankari** [comment]
as expected, the debian clang does not work on alpine.
I can see it `cat /usr/bin/clang-format` but cannot run it

```
# /usr/bin/clang-format
/bin/sh: /usr/bin/clang-format: not found
```

I really don't want to get sucked in to an alpine/debian crossport at this point.
So what can we do?
- run the clang stuff during build: but how will it have access to the layers on the runkit? Note that for PWD, and for others to  play with it, we really want to avoid having to have people check out code (although I guess it may come to that)
- run the clang stuff during runtime with a shared volume: which also sucks because then we have to wait to build everything before we run and it looks like their build process is fairly slow, given that they have switched to a large-ubuntu runner. PWD is not gonna like that, and is probably going to time out, giving us a fairly sucky experience

***First order of business once we start real work is to fix their build process...***

But for now, it seems like the best option is #3 - use their debian container

The debian `Dockerfile` seems to be maintained in sync with the alpine `Dockerfile`, *but* it is not built automatically.
Since we are doing all this manually, it seems like we should be able to build from it (potentially as part of docker-compose build using the github URL so we don't have too many copies sitting around), and then add clang stuff to in our dockerfile  and build everything.

**Comment by shankari** [comment]
Added an image that builds from the Dockerfile on GitHub and ran into 

```
#1 [internal] load git source https://github.com/EVerest/everest-ci.git#master:docker/images/build-kit
#1 ERROR: subdir not supported yet
------
 > [internal] load git source https://github.com/EVerest/everest-ci.git#master:docker/images/build-kit:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to read dockerfile: failed to load cache key: subdir not supported yet
```

This can apparently be fixed by using buildx. They use buildx for their Dockerfile, but I think that is primarily to support debian and alpine at the same time. Let's copy over their Dockerfile and see if we can at least get it to work

**Comment by shankari** [comment]
That built successfully after copying over the entryfile and the maven settings. Trying to run the scripts now...

**Comment by shankari** [comment]
Can't start because of

```
$ docker run -it everest-demo-e2e-test-server:latest /bin/sh
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/entrypoint.sh": permission denied: unknown.
```

```
$ docker run -it everest-demo-build-kit-debian:latest /bin/sh
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/entrypoint.sh": permission denied: unknown.
ERRO[0000] error waiting for container: context canceled
```

Overriding with `--entrypoint` worked

**Comment by shankari** [comment]
clang now seems to run successfully - at least it doesn't give any errors except highlighting this entry that needs to be reformatted

```
root@2bad2ccf3881:/ext# /usr/bin/run-clang-format . --extensions "hpp,cpp" -e ./cache -r
--- ./modules/OCPP201/OCPP201.cpp	(original)
+++ ./modules/OCPP201/OCPP201.cpp	(reformatted)
@@ -438,8 +438,8 @@
                     evse_id, 1, session_id, timestamp,
                     ocpp::v201::TriggerReasonEnum::RemoteStart, // FIXME(piet): Use proper reason here
                     meter_value, id_token, std::nullopt, reservation_id, remote_start_id,
-                    ocpp::v201::ChargingStateEnum::Charging); // FIXME(piet): add proper groupIdToken +
-                                                              // ChargingStateEnum
+                    ocpp::v201::ChargingStateEnum::Charging);   // FIXME(piet): add proper groupIdToken +
+                                                                // ChargingStateEnum
                 break;
             }
             case types::evse_manager::SessionEventEnum::TransactionFinished: {
```

**Comment by shankari** [comment]
After fiddling around a bit with the paths (the code needs to be in `/ext/sources`), I can get it to start compiliing by running the compile script manually (sources/.ci/compile_script). Will try to copy over the entrypoint correctly, but unclear if we need to use it.

Definitely cannot run the build every time - it is taking many many minutes and is not even done yet.

**Comment by shankari** [comment]
Given this https://github.com/shankari/everest-demo/issues/1#issuecomment-1685418895, which appears to only use the format functionality in `clang-format`, we probably don't need to run clang for the demo, which will simplify this a lot 😄 
But let's first get it working with all of their modules and then simplify

**Comment by shankari** [comment]
Built successfully!

```
[422/576] Generating ld-ev for module Example
Creating file Example/ld-ev.hpp
Creating file Example/ld-ev.cpp
[576/576] Linking CXX executable modules/examples/Example/Example
```

Let's first run the tests successfully before trying to run the SIL

```
Successfully built everestpy
Installing collected packages: everestpy
Successfully installed everestpy-0.0.4
```

e2e tests are failing because they are launched in yet another docker container, and while trying to run 

```
cd tests
pytest --everest-prefix ../dist core_tests/startup_tests.py
```

there is no `../dist`

But there is a `/workspace/dist`, which is the only directory there.  Seems to work better in the action, hopefully will be fixed once we actually get the endpoint set up correctly

```
[edm]: Scanning "/ext/source" for dependencies.
[edm]: Parsing dependencies file: /ext/source/dependencies.yaml
[edm]: Using parent directory as workspace path: /ext
[edm]: Using workspace directory "/ext".
```

Running `pytest --everest-prefix /workspace/dist core_tests/startup_tests.py` instead passes everything other than the MQTT tests, which will not work unless we run it in the docker-compose environment. Let's try to at least find the SIL scripts although we can't run them without the MQTT

**nb** `test_and_install` deletes the build. Why?!`

**Comment by shankari** [comment]
In parallel, let's get started on creating the script to launch the SIL and export the port.
The exported port should be openable from the PWD UI
https://stackoverflow.com/questions/61104342/play-with-docker-access-ip-address-on-webbrowser

**Comment by shankari** [comment]
Ok so the scripts are in the build
- `everest-core/build/run-scripts/run-sil.sh` seems to be fairly simple, and fails because of lack of MQTT.
- `everest-core/build/run-scripts/nodered-sil.sh` seems to run docker?

Yup

```
cd /workspace/dist/share/everest/docker
docker build -t everest-nodered .
docker run --rm --network host --name everest_nodered --mount type=bind,source=/ext/source/config/nodered/config-sil-flow.json,target=/data/flows.json everest-nodered
```

It builds a new image and runs it. So in a way it is simple - the currrent service can just `run-sil.sh`
and we can have another service for nodered. But it is a bit of a problem that the nodered image is built from the `dist` directory.

- Can we get it any other way?
- Given that the nodered-sil.sh doesn't seem to export any ports, it looks like all the port export is purely from `run-sil.sh`

And the only file in 

```
# ls -al /workspace/dist/share/everest/docker
total 12
drwxr-xr-x  2 root root 4096 Aug 20 23:36 .
drwxr-xr-x 10 root root 4096 Aug 20 23:36 ..
-rw-r--r--  1 root root  188 Aug 20 22:16 Dockerfile

# cat Dockerfile
FROM nodered/node-red:2.2.3
RUN npm install node-red-dashboard
RUN npm install node-red-contrib-ui-actions
RUN npm install node-red-node-ui-table
RUN npm install node-red-contrib-ui-level
```

Which is also a file here https://github.com/EVerest/everest-core/blob/bf2de55b40eb44f84072a7e2daf07f891945d185/cmake/assets/docker/Dockerfile#L3

**Comment by shankari** [comment]
There seems to be a little bit of confusion around which port to use.
Per https://github.com/EVerest/everest.github.io/blob/c8ce5904f620c39fb1880430dac033720fc087b2/docs/contacts-snapshot/_sources/tutorials/run_sil/index.rst.txt#L32 the nodered dashboard and the UI are both at port 1880.

But per https://github.com/EVerest/everest.github.io/blob/c8ce5904f620c39fb1880430dac033720fc087b2/docs/latest/_sources/general/03_quick_start_guide.rst.txt#L6, updated more recently overall, the admin dashboard (which appears to be the nodered dashboard) is at 8849

**Comment by shankari** [comment]
The nodered dashboard is essentially independent and doesn't really need the everest build as long as it has all the configs. Let's try to run it independently and see which ports it exposes.

- node-red runs on port 1880 by default https://nodered.org/docs/getting-started/docker

Launch succeeds, but it is not able to run MQTT stuff. 

```

> node-red-docker@2.2.3 start /usr/src/node-red
> node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"

21 Aug 00:42:52 - [info]

Welcome to Node-RED
===================

21 Aug 00:42:52 - [info] Node-RED version: v2.2.3
21 Aug 00:42:52 - [info] Node.js  version: v14.18.2
21 Aug 00:42:52 - [info] Linux 5.15.49-linuxkit x64 LE
21 Aug 00:42:52 - [info] Loading palette nodes
21 Aug 00:42:52 - [info] Dashboard version 3.5.0 started at /ui
21 Aug 00:42:52 - [info] Settings file  : /data/settings.js
21 Aug 00:42:52 - [info] Context store  : 'default' [module=memory]
21 Aug 00:42:52 - [info] User directory : /data
21 Aug 00:42:52 - [warn] Projects disabled : editorTheme.projects.enabled=false
21 Aug 00:42:52 - [info] Flows file     : /data/flows.json
21 Aug 00:42:52 - [info] Server now running at http://127.0.0.1:1880/
21 Aug 00:42:52 - [warn]

21 Aug 00:42:53 - [info] [mqtt-broker:fc8686af.48d178] Connection failed to broker: mqtt://localhost:1883
21 Aug 00:43:08 - [info] [mqtt-broker:fc8686af.48d178] Connection failed to broker: mqtt://localhost:1883
21 Aug 00:43:23 - [info] [mqtt-broker:fc8686af.48d178] Connection failed to broker: mqtt://localhost:1883
21 Aug 00:43:38 - [info] [mqtt-broker:fc8686af.48d178] Connection failed to broker: mqtt://localhost:1883
```

Cannot also access the port, but that is likely due to the port not being mapped. Trying to re-run with the port mapping...
Still not working, likely due to the MQTT. Presumably we don't need the port mapping due to the --host?

**Comment by shankari** [comment]
This works:
`docker run --rm -it -p 1880:1880 -v node_red_data:/data --name mynodered nodered/node-red:2.2.3`

this doesn't work

`docker run --rm -it --network host -v node_red_data:/data --name mynodered nodered/node-red:2.2.3`

It seems to start up fine, but the page is not accessible.

```
21 Aug 00:58:40 - [info] Node-RED version: v2.2.3
21 Aug 00:58:40 - [info] Node.js  version: v14.18.2
21 Aug 00:58:40 - [info] Linux 5.15.49-linuxkit x64 LE
21 Aug 00:58:41 - [info] Loading palette nodes
21 Aug 00:58:41 - [info] Settings file  : /data/settings.js
21 Aug 00:58:41 - [info] Context store  : 'default' [module=memory]
21 Aug 00:58:41 - [info] User directory : /data
21 Aug 00:58:41 - [warn] Projects disabled : editorTheme.projects.enabled=false
21 Aug 00:58:41 - [info] Flows file     : /data/flows.json
21 Aug 00:58:41 - [warn]

---------------------------------------------------------------------
Your flow credentials file is encrypted using a system-generated key.

If the system-generated key is lost for any reason, your credentials
file will not be recoverable, you will have to delete it and re-enter
your credentials.

You should set your own key using the 'credentialSecret' option in
your settings file. Node-RED will then re-encrypt your credentials
file using your chosen key the next time you deploy a change.
---------------------------------------------------------------------

21 Aug 00:58:41 - [warn] Encrypted credentials not found
21 Aug 00:58:41 - [info] Server now running at http://127.0.0.1:1880/
21 Aug 00:58:41 - [info] Starting flows
21 Aug 00:58:41 - [info] Started flows
```

Checked the container

```
            "ExposedPorts": {
                "1880/tcp": {}
            },

            "Ports": {},
```

Ok... so I don't know how the docker command was ever supposed to work.
Now running the working command with the mounted flow...

```
docker run --rm -it -p 1880:1880 -v node_red_data:/data  --mount type=bind,source=/.../everest-core/config/nodered/config-sil-flow.json,target=/data/flows.json --name mynodered nodered/node-red:2.2.3
```

works but has errors with the load

<img width="505" alt="Screenshot 2023-08-20 at 6 05 16 PM" src="https://github.com/shankari/everest-demo/assets/2423263/87f82094-33b9-4d82-80be-56ed5bbf2da3">

**Comment by shankari** [comment]
Building from the everest-nodered image works, and removes the errors, but a lot of the external commands are having trouble connecting, likely due to the MQTT issue

```
$ docker run --rm -it -p 1880:1880 -v node_red_data:/data  --mount type=bind,source=/.../everest-core/config/nodered/config-sil-flow.json,target=/data/flows.json --name mynodered everest-nodered
```

<img width="910" alt="Screenshot 2023-08-20 at 6 07 50 PM" src="https://github.com/shankari/everest-demo/assets/2423263/d1df919e-d88c-49b7-8a9a-be633a430202">

**Comment by shankari** [comment]
- It seems like this is not strictly required for the operation of the system, but it is great for a demo.
- It seems like it will do very well once we get the MQTT to work. But where do we configure the MQTT and which environment variable do we set?

Ahah!
https://cookbook.nodered.org/mqtt/connect-to-broker

No environment variable

it is hardcoded in the flow file - e.g. 

```
        "type": "mqtt-broker",
        "name": "",
        "broker": "localhost",
        "port": "1883",
        "clientid": "",
        "usetls": false,
```

Can confirm that this is the right way to do it "Linking Containers". Not sure why they don't just use `docker-compose`

On production, we will copy the configs into the container instead of mounting them, so can modify the hostname.
On dev, there's no point in doing that then we will change the mounted values, and we might as well change them ahead of time if needed in that case.

So let's copy over the flow files to our demo directory and manually rename them.

There are other components:
- STeVE (which is something to do with OCPP)
- Admin Panel, which is indeed separate (https://github.com/EVerest/everest-admin-panel)

**Comment by shankari** [comment]
Note also that it looks like admin-panel is used in the EVerest framework
https://github.com/EVerest/everest-framework/blob/df044f9a4d723a3e7f9ff9fe966c645b9b33c2c3/src/controller/CMakeLists.txt#L47

But neither of them seem to be in the source

```
/ext/source# grep -r \*everestjs\* .
/ext/source#

/ext/source# grep -r \*EvModuleList\* .
/ext/source#
```

So we might want to deal an expanded dockerfile that follows all the instructions below:
https://everest.github.io/nightly/general/03_quick_start_guide.html

**Comment by shankari** [comment]
We cannot use the clang in the Dockerfile - if there is any super small error, it fails

```
#13 [ 8/13] RUN /usr/bin/run-clang-format . --extensions "hpp,cpp" -e ./cache -r
#13 0.236 --- ./run-clang-format/src/third_party/qux.cpp	(original)
#13 0.236 +++ ./run-clang-format/src/third_party/qux.cpp	(reformatted)
#13 0.236 @@ -1,8 +1,8 @@
#13 0.236  // This code is ignored by the .clang-format-ignore file.
#13 0.236
#13 0.236 -int qux (bool cond)  {
#13 0.236 -  if(cond){
#13 0.236 -    return - 1;
#13 0.236 +int qux(bool cond) {
#13 0.236 +  if (cond) {
#13 0.236 +    return -1;
#13 0.236    }
#13 0.236
#13 0.236    return 0;
#13 0.236 --- ./run-clang-format/src/foo.cpp	(original)
#13 0.236 +++ ./run-clang-format/src/foo.cpp	(reformatted)
#13 0.236 @@ -2,6 +2,6 @@
#13 0.236
#13 0.236  std::string foo(const std::string &a, const std::string &b) {
#13 0.236    std::string sum;
#13 0.236 -   sum = a + b;
#13 0.236 +  sum = a + b;
#13 0.236    return sum;
#13 0.236  }
#13 ERROR: executor failed running [/bin/sh -c /usr/bin/run-clang-format . --extensions "hpp,cpp" -e ./cache -r]: exit code: 1
------
 > [ 8/13] RUN /usr/bin/run-clang-format . --extensions "hpp,cpp" -e ./cache -r:
#13 0.236 --- ./run-clang-format/src/foo.cpp	(original)
#13 0.236 +++ ./run-clang-format/src/foo.cpp	(reformatted)
#13 0.236 @@ -2,6 +2,6 @@
#13 0.236
#13 0.236  std::string foo(const std::string &a, const std::string &b) {
#13 0.236    std::string sum;
#13 0.236 -   sum = a + b;
#13 0.236 +  sum = a + b;
#13 0.236    return sum;
#13 0.236  }
------
failed to solve: executor failed running [/bin/sh -c /usr/bin/run-clang-format . --extensions "hpp,cpp" -e ./cache -r]: exit code: 1
```

**Comment by shankari** [comment]
Compile is also failing, running it manually first and then putting it into the `Dockerfile`

```
#16 [11/12] RUN run-script compile
#16 0.249 /bin/sh: 1: run-script: not found
#16 ERROR: executor failed running [/bin/sh -c run-script compile]: exit code: 127
------
 > [11/12] RUN run-script compile:
#16 0.249 /bin/sh: 1: run-script: not found
------
failed to solve: executor failed running [/bin/sh -c run-script compile]: exit code: 127
```

**Comment by shankari** [comment]
Entrypoint is still broken

```
$ docker run -it everest-demo-manager /bin/bash
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/entrypoint.sh": permission denied: unknown.
ERRO[0000] error waiting for container: context canceled
```

**Comment by shankari** [comment]
It was just the permissions - `bash /entrypoint.sh run-script compile` works
Given that we can't run clang anyway, I am very tempted to just go back to the alpine build, which will make our life much easier overall

**Comment by shankari** [comment]
After switching to alpine, the image was built successfully

```
#12 exporting to image
#12 exporting layers
#12 exporting layers 14.0s done
#12 writing image sha256:9bf3c9bb58a2774b678dac40ef2e91f42a114f03490bdff5e1d1d223b22bc105 done
#12 naming to docker.io/library/everest-demo-manager 0.0s done
#12 DONE 14.0s
```

Now I will exec into the manager and try to run

**Comment by shankari** [comment]
Entrypoint in the dockerfile is not working?! but is in the docker-compose
The nodered container is now connected to MQTT and shows as "connected"

```
everest-demo-mqtt-server-1       | 1692590573: New connection from 192.168.2.165:42672 on port 1883.
everest-demo-mqtt-server-1       | 1692590573: New client connected from 192.168.2.165:42672 as nodered_ad70af80b46de7b1 (p2, c1, k60).
everest-demo-node-red-1          | 21 Aug 04:02:53 - [info] [mqtt-broker:fc8686af.48d178] Connected to broker: mqtt://mqtt-server:1883
```
<img width="717" alt="Screenshot 2023-08-20 at 9 05 22 PM" src="https://github.com/shankari/everest-demo/assets/2423263/69c0dac0-b6de-4950-a5a2-b9a8bfcf5e68">

Start up the manager as well, and that seems to work...

```
2023-08-21 04:05:58.882189 [INFO] manager          :: EVerest manager starting using /ext/source/config/config-sil.yaml
2023-08-21 04:05:58.882255 [INFO] manager          :: EVerest using MQTT broker mqtt-server:1883
2023-08-21 04:05:58.882280 [INFO] manager          :: EVerest telemetry enabled
2023-08-21 04:05:59.355271 [INFO] everest_ctrl     :: everest controller process started ...
2023-08-21 04:05:59.355498 [INFO] everest_ctrl     :: Launching controller service on port 8849
2023-08-21 04:06:00.996474 [INFO] api:API          :: Module api initialized.
2023-08-21 04:06:01.147806 [INFO] connector_1:Evs  :: Module connector_1 initialized.
2023-08-21 04:06:01.151094 [INFO] energy_manager:  :: Module energy_manager initialized.
2023-08-21 04:06:01.281156 [INFO] token_validator  :: Module token_validator initialized.
2023-08-21 04:06:01.285056 [INFO] grid_connection  :: Module grid_connection_point initialized.
2023-08-21 04:06:01.322436 [INFO] auth:Auth        :: Module auth initialized.

2023-08-21 04:06:01.389648 [INFO] iso15118_charge  :: TCP server on eth0 is listening on port [fe80::42:c0ff:fea8:2a4%275]:61341

2023-08-21 04:06:01.389757 [INFO] iso15118_charge  :: TLS server on eth0 is listening on port [fe80::42:c0ff:fea8:2a4%275]:64109

2023-08-21 04:06:01.389800 [INFO] iso15118_charge  :: SDP socket setup succeeded
2023-08-21 04:06:01.389941 [INFO] iso15118_charge  :: Module iso15118_charger initialized.
2023-08-21 04:06:01.435038 [INFO] car_simulator:J  :: Module car_simulator initialized.
2023-08-21 04:06:01.530832 [INFO] persistent_stor  :: Module persistent_store initialized.
2023-08-21 04:06:01.531557 [INFO] setup:Setup      :: Module setup initialized.
2023-08-21 04:06:01.582444 [INFO] connector_1_pow  :: Module connector_1_powerpath initialized.
2023-08-21 04:06:01.596913 [INFO] slac:JsSlacSimu  :: Module slac initialized.
2023-08-21 04:06:01.729395 [INFO] token_provider:  :: Module token_provider initialized.
2023-08-21 04:06:02.206653 [INFO] iso15118_car     :: Module iso15118_car initialized.
2023-08-21 04:06:02.227318 [INFO] manager          :: >>> All modules are initialized. EVerest up and running <<<
2023-08-21 04:06:03.673677 [INFO] connector_1:Evs  :: Max AC hardware capabilities: 32A/3ph
2023-08-21 04:06:03.899127 [INFO] connector_1:Evs  :: AC HLC mode enabled.
2023-08-21 04:06:03.899773 [INFO] connector_1:Evs  :: 🌀🌀🌀 Ready to start charging 🌀🌀🌀
```


| UI works | charging works | Error is detected| 
|-----|-----|----|
| <img width="341" alt="Screenshot 2023-08-20 at 9 07 45 PM" src="https://github.com/shankari/everest-demo/assets/2423263/7c992851-be78-426a-9361-a23858b7cfdb"> | <img width="341" alt="Screenshot 2023-08-20 at 9 10 27 PM" src="https://github.com/shankari/everest-demo/assets/2423263/30ab8d14-68c8-4bcf-a9d5-17e88cf8e599"> | <img width="341" alt="Screenshot 2023-08-20 at 9 09 53 PM" src="https://github.com/shankari/everest-demo/assets/2423263/05cb3823-ea7d-44b1-a418-b8b639d611f5"> |

**Comment by shankari** [comment]
ok so the goal for tonight is to clean this up and put it into PWD.

Slight change of plan:
- given the time taken to build, the docker-compose I'm working on will be a docker-compose.build.yml
- it will still copy all the files into the containers and tag them appropriately
- the docker-compose.pwd.yml will just launch directly from the images - no build, no slowness
- note that the "Try with PWD" button is currently broken but we can still put in instructions to run with docker-compose in PWD and they should be similar
https://github.com/play-with-docker/play-with-docker/issues/335

**Comment by shankari** [comment]
I think that these are their official docker instructions:
https://github.com/EVerest/everest-utils/tree/main/docker

It includes a debian based "everest-playground" 
https://github.com/EVerest/everest-utils/blob/main/docker/everest-docker-image/Dockerfile

And it doesn't do clang 😄 but it is not tested by the CI so I don't trust it as much
They also want to seem to create a dfiferent image for each configuration, which is cool but should definitely be automated.

But more importantly, it also includes StEVe instructions so I can get OCPP to work.

**Comment by shankari** [comment]
Using StEVe is a giant pain as well. There is no pre-built image - you have to check out the code and build from scratch.
And they are using `dockerize` instead of `docker-compose` to wait for the DB instance to start up
And it's not clear how to pass in the URL to the mariadb, or is it assumed to be `localhost`?

**Comment by shankari** [comment]
High level conclusion:
- functionality seems to be fairly good
- needs lots and lots of software engineering and tooling support

For example:
- why is the clang failing in the build when it is supposed to be part of the CI?
- why are there two debian dockerfiles in two locations?
- why isn't there is a single docker-compose that launches everything without needing to build?
- should we have different images for each config? if so, what is the CI to generate them?
- can we have a stripped down release version that doesn't need everything in dist and that doesn't use up 230 MB
- ...

In other words, like a typical underfunded open-source project. I happen to be very familiar with at least one of those.

I think that they have a strong technical foundation to build on and we can quickly make the engineering and tooling improvements needed, by staffing up with strong software engineers, even if they don't have specific domain expertise. That can free them (and other industry stakeholders) up to do more of the domain specific work.

**Comment by shankari** [comment]
Note that packages are published as private by default. It looks also that we can't change this for user repos, but can do so at an org level. Should make sure that this is turned on once we contribute this to the org repo

**Comment by shankari** [comment]
Quick notes about steve.
- only supports upto OCPP 1.6
   - need to check with EVerest on their plans for OCPP 2.0.1 at the charging network layer
- does not appear to be super active recently, although they have 301 forks and 644 stars, so presumably somebody is still using it
- GPL, which is going to be a dealbreaker unless they are willing to change their license

**Comment by shankari** [comment]
Looks like steve doesn't build?

```
Downloaded from central: https://repo1.maven.org/maven2/org/apache/apache/23/apache-23.pom (18 kB at 91 kB/s)
everest-demo-steve-1        | [INFO]
everest-demo-steve-1        | [INFO] -------------------------< de.rwth.idsg:steve >-------------------------
everest-demo-steve-1        | [INFO] Building steve 3.4.5
everest-demo-steve-1        | [INFO] --------------------------------[ jar ]---------------------------------
everest-demo-steve-1        | Downloading from central: https://repo.maven.apache.org/maven2/com/mycila/license-maven-plugin/4.1/license-maven-plugin-4.1.pom
everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [INFO] BUILD FAILURE
everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [INFO] Total time:  3.698 s
everest-demo-steve-1        | [INFO] Finished at: 2023-08-23T02:32:21Z
everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [ERROR] Plugin com.mycila:license-maven-plugin:4.1 or one of its dependencies could not be resolved: Failed to read artifact descriptor for com.mycila:license-maven-plugin:jar:4.1: Could not transfer artifact com.mycila:license-maven-plugin:pom:4.1 from/to central (https://repo.maven.apache.org/maven2): PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target -> [Help 1]
```

**Comment by shankari** [comment]
ok so there's something kind of weird here. the steve dockerfile (https://github.com/EVerest/everest-utils/tree/main/docker/steve) builds off `maven:3.6.1-jdk-11`. When I tried to run the base image directly, I didn't specify the shell, so I got the errror below. Doh!

```
Status: Downloaded newer image for maven:3.6.1-jdk-11
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.084 s
[INFO] Finished at: 2023-08-23T02:40:05Z
[INFO] ------------------------------------------------------------------------
[ERROR] No goals have been specified for this build. You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
```

On using the shell, and `mvn package` I got the same error as above. Looking more carefully, it is the stupid ssl error...
After resolving that, it builds, but then errors out while trying to connect to mariadb. So pretty close.

BUT the build process doesn't actually build. It only downloads the data. 

```
$ docker-compose -f docker-compose.run-sil-ocpp.build.yml build --no-cache --pull --progress=plain
```

The build happens when you launch, because I guess the mvn image installs the `pom.xml` in the workdir.

**Comment by shankari** [comment]
On launch, the build succeeded, we connected to the database, and then appeared to start a bunch of migrations

```
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.6.8 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.6.9 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.0 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.1 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.2 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.3 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.6 - update"
everest-demo-steve-1        | [INFO] Migrating schema `ocpp-db` to version "0.7.7 - update"
everest-demo-steve-1        | [ERROR] Migration of schema `ocpp-db` to version "0.7.7 - update" failed! Please restore backups and roll back database and code!everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [INFO] BUILD FAILURE
everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [INFO] Total time:  35.684 s
everest-demo-steve-1        | [INFO] Finished at: 2023-08-23T03:39:14Z
everest-demo-steve-1        | [INFO] ------------------------------------------------------------------------
everest-demo-steve-1        | [ERROR] Failed to execute goal org.flywaydb:flyway-maven-plugin:7.9.1:migrate (default) on project steve: org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException:
everest-demo-steve-1        | [ERROR] Migration V0_7_7__update.sql failed
everest-demo-steve-1        | [ERROR] -----------------------------------
everest-demo-steve-1        | [ERROR] SQL State  : HY000
everest-demo-steve-1        | [ERROR] Error Code : 1833
everest-demo-steve-1        | [ERROR] Message    : Cannot change column 'chargeBoxId': used in a foreign key constraint 'ocpp@002ddb/FK_chargeBoxId_c' of table 'ocpp@002ddb/connector'
everest-demo-steve-1        | [ERROR] Location   : /steve/src/main/resources/db/migration/V0_7_7__update.sql (/steve/src/main/resources/db/migration/V0_7_7__update.sql)
everest-demo-steve-1        | [ERROR] Line       : 1
```

We are setting this up from scratch - do we really need all these migrations?

**Comment by shankari** [comment]
Known issue <a href="https://github.com/steve-community/steve/pull/1213/files">https://github.com/steve-community/steve/pull/1213/files</a>, applying recommended fix

**Comment by shankari** [comment]
That starts it up

```
everest-demo-steve-1        | . Done!
everest-demo-steve-1        | Hint: You can stop the application by pressing CTRL+C
everest-demo-steve-1        |
everest-demo-steve-1        | Access the web interface using
everest-demo-steve-1        | - http://192.168.2.197:8180/steve/manager/home
everest-demo-steve-1        | - https://192.168.2.197:8443/steve/manager/home
everest-demo-steve-1        | SOAP endpoint for OCPP
everest-demo-steve-1        | - http://192.168.2.197:8180/steve/services/CentralSystemService
everest-demo-steve-1        | - https://192.168.2.197:8443/steve/services/CentralSystemService
everest-demo-steve-1        | WebSocket/JSON endpoint for OCPP
everest-demo-steve-1        | - ws://192.168.2.197:8180/steve/websocket/CentralSystemService/(chargeBoxId)
everest-demo-steve-1        | - wss://192.168.2.197:8443/steve/websocket/CentralSystemService/(chargeBoxId)
```

The manager is still not able to connect to it, but we just need to override the config, I think

**Comment by shankari** [comment]
- I can't find a template config anywhere, so copying the config locally and then copying it to the correct location 
- the build process takes a *long* time, and the build process appears to want to connect to the DB and migrate at the same time. The migration is apparently automatic by using the flyway maven plugin:migrate task. Do we really need to migrate while building?

We can spin this as "resilience to failure" etc but would like to see if we don't have to take quite as long.

**Comment by shankari** [comment]
Apparently the "needs DB to compile" issue is known <a href="https://github.com/steve-community/steve/issues/420">https://github.com/steve-community/steve/issues/420</a> and is being considered for improvement, but has not yet been improved and might not be improved in the near future. I certainly don't have time to fix it now  <a href="https://github.com/steve-community/steve/issues/320">and the maintainers are pretty opposed to containerization and the tone is pretty unpleasant</a>

**Comment by shankari** [comment]
so now the connection starts

```
2023-08-23 04:49:14.567827 [INFO] ocpp:OCPP        :: Connecting to plain websocket at uri: ws://steve:8180/steve/websocket/CentralSystemService/cp001 with profile: 0
```

and is received but fails because there is no chargepoint ID

```
everest-demo-steve-1        | [WARN ] 2023-08-23 04:49:26,819 org.eclipse.jetty.server.HttpChannel (qtp247162961-16) - /steve/websocket/CentralSystemService/cp001
everest-demo-steve-1        | org.springframework.web.util.NestedServletException: Request processing failed; nested exception is org.springframework.web.socket.server.HandshakeFailureException: ChargeBoxId 'cp001' is not recognized.
everest-demo-steve-1        | 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-5.3.7.jar:5.3.7]
```

let's try to open the admin dashboard and see if we can create the point

**Comment by shankari** [comment]
After manually adding a chargeBox called `cp001` using the blue "Add New" button, the connection worked!

```
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,452 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-29) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Connection is established
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,467 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-27) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Received: [2,"cc7ba7d6-66a9-47e3-a3be-91a5fae49cdf","BootNotification",{"chargeBoxSerialNumber":"cp001","chargePointModel":"Yeti","chargePointVendor":"Pionix","firmwareVersion":"0.1"}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,489 de.rwth.idsg.steve.service.CentralSystemService16_Service (qtp247162961-27) - The boot of the chargebox 'cp001' with registration status 'Optional[ACCEPTED]' is acknowledged.
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,585 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-27) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Sending: [3,"cc7ba7d6-66a9-47e3-a3be-91a5fae49cdf",{"status":"Accepted","currentTime":"2023-08-23T05:14:05.489Z","interval":14400}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,593 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-43) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Received: [2,"938ca1d8-f39a-4184-85b4-6f3029b026b9","StatusNotification",{"connectorId":0,"errorCode":"NoError","status":"Available"}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,620 de.rwth.idsg.steve.repository.impl.OcppServerRepositoryImpl (qtp247162961-43) - The connector cp001/0 is NEW, and inserted into DB.
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,628 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-43) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Sending: [3,"938ca1d8-f39a-4184-85b4-6f3029b026b9",{}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:14:05,632 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-27) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Received: [2,"fee091f9-da9c-4733-b08c-71dff605dd18","StatusNotification",{"connectorId":1,"errorCode":"NoError","status":"Available"}]
``` 

and 

```
2023-08-23 05:14:05.390127 [INFO] ocpp:OCPP        :: Reconnecting to plain websocket at uri: ws://steve:8180/steve/websocket/CentralSystemService/cp001 with profile: 0
2023-08-23 05:14:05.446145 [INFO] ocpp:OCPP        :: OCPP client successfully connected to plain websocket server
2023-08-23 05:15:05.595665 [INFO] ocpp:OCPP        :: Checking if OCSP cache should be updated
2023-08-23 05:15:05.596462 [INFO] ocpp:OCPP        :: Requesting OCSP response.
2023-08-23 05:15:05.600287 [INFO] ocpp:OCPP        :: Checking if V2GCertificate has expired
2023-08-23 05:15:05.600821 [INFO] ocpp:OCPP        :: V2GCertificate is invalid in 0 days. Requesting new certificate with certificate signing request
```

**Comment by shankari** [comment]
If I try to plugin a car using the standard UI, however, I get the error:

```
2023-08-23 05:17:28.938506 [INFO] evse_manager_1:  :: SYS  Session logging started.
2023-08-23 05:17:28.938728 [INFO] evse_manager_1:  :: EVSE IEC Session Started: EVConnected
2023-08-23 05:17:28.989786 [INFO] ocpp:OCPP        :: Logging OCPP messages to html file: /tmp/everest-logs/2023-08-23T05:17:28.938Z-51567328-285c-4ce3-a679-8987a246484f/incomplete-ocpp.html
2023-08-23 05:17:30.301639 [ERRO] energy_manager: std::vector<types::energy::EnforcedLimits> module::EnergyManager::run_optimizer(types::energy::EnergyFlowRequest) :: Trading: Maximum number of trading rounds reached.
2023-08-23 05:17:32.012429 [ERRO] energy_manager: std::vector<types::energy::EnforcedLimits> module::EnergyManager::run_optimizer(types::energy::EnergyFlowRequest) :: Trading: Maximum number of trading rounds reached.
2023-08-23 05:17:33.708568 [ERRO] energy_manager: std::vector<types::energy::EnforcedLimits> module::EnergyManager::run_optimizer(types::energy::EnergyFlowRequest) :: Trading: Maximum number of trading rounds reached.
2023-08-23 05:17:35.264066 [ERRO] energy_manager: std::vector<types::energy::EnforcedLimits> module::EnergyManager::run_optimizer(types::energy::EnergyFlowRequest) :: Trading: Maximum number of trading rounds reached.
2023-08-23 05:17:36.903633 [ERRO] energy_manager: std::vector<types::energy::EnforcedLimits> module::EnergyManager::run_optimizer(types::energy::EnergyFlowRequest) :: Trading: Maximum number of trading rounds reached.
```

And there are some promising-looking messages on the charging network side

```
everest-demo-steve-1        | [INFO ] 2023-08-23 05:15:05,608 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-36) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Received: [2,"ab6930f6-bf8d-477d-9c1e-d9befb11e239","DataTransfer",{"data":"{\"certificateType\":\"V2GCertificate\",\"csr\":\"-----BEGIN CERTIFICATE REQUEST-----\\nMIIBKjCB0QIBADBDMQswCQYDVQQGEwJERTEPMA0GA1UECgwGUGlvbml4MQ4wDAYD\\nVQQDDAVjcDAwMTETMBEGCgmSJomT8ixkARkWA0NQTzBZMBMGByqGSM49AgEGCCqG\\nSM49AwEHA0IABA1ax+CTmpQuDa46+uPqWvSq0Eh0Jl6a1G7K4bUVtHogCYr+GuOb\\nbrkvjd5ZpuNbpDhheUQ15U7ih/5LC6cUUISgLDAqBgkqhkiG9w0BCQ4xHTAbMAsG\\nA1UdDwQEAwIDiDAMBgNVHRMBAf8EAjAAMAoGCCqGSM49BAMCA0gAMEUCIAwi8oUK\\nYfUVdflSSs53+57PHrDxV6ot4n6GuChfB61yAiEAqjK1EkIpY5ARU2M5RRB/zJ2K\\n9OaW5J2mVzfEk8Bfi6A=\\n-----END CERTIFICATE REQUEST-----\\n\"}","messageId":"SignCertificate","vendorId":"org.openchargealliance.iso15118pnc"}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:15:05,609 de.rwth.idsg.steve.service.CentralSystemService16_Service (qtp247162961-36) - [Data Transfer] Charge point: cp001, Vendor Id: org.openchargealliance.iso15118pnc
everest-demo-steve-1        | [INFO ] 2023-08-23 05:15:05,609 de.rwth.idsg.steve.service.CentralSystemService16_Service (qtp247162961-36) - [Data Transfer] Message Id: SignCertificate
everest-demo-steve-1        | [INFO ] 2023-08-23 05:15:05,610 de.rwth.idsg.steve.service.CentralSystemService16_Service (qtp247162961-36) - [Data Transfer] Data: {"certificateType":"V2GCertificate","csr":"-----BEGIN CERTIFICATE REQUEST-----\nMIIBKjCB0QIBADBDMQswCQYDVQQGEwJERTEPMA0GA1UECgwGUGlvbml4MQ4wDAYD\nVQQDDAVjcDAwMTETMBEGCgmSJomT8ixkARkWA0NQTzBZMBMGByqGSM49AgEGCCqG\nSM49AwEHA0IABA1ax+CTmpQuDa46+uPqWvSq0Eh0Jl6a1G7K4bUVtHogCYr+GuOb\nbrkvjd5ZpuNbpDhheUQ15U7ih/5LC6cUUISgLDAqBgkqhkiG9w0BCQ4xHTAbMAsG\nA1UdDwQEAwIDiDAMBgNVHRMBAf8EAjAAMAoGCCqGSM49BAMCA0gAMEUCIAwi8oUK\nYfUVdflSSs53+57PHrDxV6ot4n6GuChfB61yAiEAqjK1EkIpY5ARU2M5RRB/zJ2K\n9OaW5J2mVzfEk8Bfi6A=\n-----END CERTIFICATE REQUEST-----\n"}
everest-demo-steve-1        | [INFO ] 2023-08-23 05:15:05,611 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-36) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Sending: [3,"ab6930f6-bf8d-477d-9c1e-d9befb11e239",{"status":"Accepted"}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:17:28,991 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-27) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Received: [2,"1bc50919-858c-49ad-9907-9a788039f6e3","StatusNotification",{"connectorId":1,"errorCode":"NoError","status":"Preparing"}]
everest-demo-steve-1        | [INFO ] 2023-08-23 05:17:29,008 de.rwth.idsg.steve.ocpp.ws.WebSocketLogger (qtp247162961-27) - [chargeBoxId=cp001, sessionId=3df7f3bc-e076-d275-090f-f2a6eb9d02ac] Sending: [3,"1bc50919-858c-49ad-9907-9a788039f6e3",{}]
```

This is almost certainly some kind of config issue, but at least it sorta works.
Will rebuild everything, re-push and update the demo README

**Comment by shankari** [comment]
> Apparently the "needs DB to compile" issue is known but has not yet been improved and might not be improved in the near future.

I had been noodling on workarounds to this because I really didn't want people to wait for 3 minutes to compile and go to the DB and create a new chargerbox....

And it came to me overnight - what we really need is to take a snapshot of the image after all the setup is done - e.g. 
https://docs.docker.com/engine/reference/commandline/commit/

**Comment by shankari** [comment]
Tried this, but it didn't work. When we started up with the committed images, steve was not able to find the required tables in the DB.

```
Invocation of init method failed; nested exception is org.jooq.exception.DataAccessException: SQL [select `ocpp-db`.`settings`.`app_id`, `ocpp-db`.`settings`.`heartbeat_interval_in_seconds`, `ocpp-db`.`settings`.`hours_to_expire`, `ocpp-db`.`settings`.`mail_enabled`, `ocpp-db`.`settings`.`mail_host`, `ocpp-db`.`settings`.`mail_username`, `ocpp-db`.`settings`.`mail_password`, `ocpp-db`.`settings`.`mail_from`, `ocpp-db`.`settings`.`mail_protocol`, `ocpp-db`.`settings`.`mail_port`, `ocpp-db`.`settings`.`mail_recipients`, `ocpp-db`.`settings`.`notification_features` from `ocpp-db`.`settings` where `ocpp-db`.`settings`.`app_id` = ?]; Table 'ocpp-db.settings' doesn't exist
```

I do see this message from the `ocpp-db`

```
2023-08-23 17:07:38+00:00 [Note] [Entrypoint]: Initializing database files


PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
To do so, start the server, then issue the following command:

'/usr/bin/mysql_secure_installation'

which will also give you the option of removing the test
databases and anonymous user created by default.  This is
strongly recommended for production servers.
```

And there's some kind of temporary DB that is set up

```
2023-08-23 17:07:40+00:00 [Note] [Entrypoint]: Starting temporary server
2023-08-23 17:07:42+00:00 [Note] [Entrypoint]: Securing system users (equivalent to running mysql_secure_installation)

2023-08-23 17:07:42+00:00 [Note] [Entrypoint]: Stopping temporary server
2023-08-23 17:07:42 0 [Note] mysqld (initiated by: unknown): Normal shutdown
2023-08-23 17:07:42 0 [Note] Event Scheduler: Purging the queue. 0 events
2023-08-23 17:07:42 0 [Note] InnoDB: FTS optimize thread exiting.
2023-08-23 17:07:42 0 [Note] InnoDB: Starting shutdown...
2023-08-23 17:07:42 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
2023-08-23 17:07:42 0 [Note] InnoDB: Buffer pool(s) dump completed at 230823 17:07:42
2023-08-23 17:07:44 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2023-08-23 17:07:44 0 [Note] InnoDB: Shutdown completed; log sequence number 60952; transaction id 21
2023-08-23 17:07:44 0 [Note] mysqld: Shutdown complete

2023-08-23 17:07:44+00:00 [Note] [Entrypoint]: Temporary server stopped

2023-08-23 17:07:44 0 [Note] mysqld: ready for connections.
Version: '10.4.30-MariaDB-1:10.4.30+maria~ubu2004'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
2023-08-23 17:07:45 8 [Warning] Aborted connection 8 to db: 'unconnected' user: 'unauthenticated' host: '192.168.2.181' (This connection closed normally without authentication)
```

I do note that we have

```
  ocpp-db:
    image: mariadb:10.4.30
    environment:
      MYSQL_RANDOM_ROOT_PASSWORD: "yes"
```

Checking to see if that is the reason and we can fix it.

**Comment by shankari** [comment]
Confirmed that tables are empty

```

MariaDB [(none)]> use ocpp-db;
Database changed
MariaDB [ocpp-db]> show tables;
Empty set (0.000 sec)
```

**Comment by shankari** [comment]
Bingo
https://stackoverflow.com/questions/44848220/how-to-commit-changed-data-in-a-mysql-container-to-a-new-image#44848440

**Comment by shankari** [comment]
Seems to be working...

```
root@af1d01e8992c:/# ls -al /var/lib/mysql
total 8
drwxr-xr-x 2 mysql mysql 4096 Aug  3 03:51 .
drwxr-xr-x 1 root  root  4096 Aug 24 04:38 ..
root@af1d01e8992c:/# ls -al /var/lib/mysql-no-volume/
total 140292
drwxr-xr-x 5 mysql mysql     4096 Aug 24 04:38 .
drwxr-xr-x 1 root  root      4096 Aug 24 04:38 ..
-rw------- 1 mysql mysql      135 Aug 24 04:38 .my-healthcheck.cnf
-rw-rw---- 1 mysql mysql 17793024 Aug 24 04:39 aria_log.00000001
-rw-rw---- 1 mysql mysql       52 Aug 24 04:38 aria_log_control
-rw-rw---- 1 mysql mysql      930 Aug 24 04:38 ib_buffer_pool
-rw-rw---- 1 mysql mysql 50331648 Aug 24 04:39 ib_logfile0
-rw-rw---- 1 mysql mysql 50331648 Aug 24 04:38 ib_logfile1
-rw-rw---- 1 mysql mysql 12582912 Aug 24 04:39 ibdata1
-rw-rw---- 1 mysql mysql 12582912 Aug 24 04:38 ibtmp1
-rw-rw---- 1 mysql mysql        0 Aug 24 04:38 multi-master.info
drwx------ 2 mysql mysql     4096 Aug 24 04:38 mysql
-rw-r--r-- 1 mysql mysql       15 Aug 24 04:38 mysql_upgrade_info
drwx------ 2 mysql mysql     4096 Aug 24 04:39 ocpp@002ddb
drwx------ 2 mysql mysql     4096 Aug 24 04:38 performance_schema
```
